src/OrleansRuntime/Development/InMemoryLeaseProvider.cs

```diff
@@ -0,0 +1,121 @@
+
+        public Task<AcquireLeaseResult[]> Acquire(string category, LeaseRequest[] leaseRequests)
+        {
+            return Task.FromResult(leaseRequests.Select(request => Acquire(category, request)).ToArray());
+        }
+
+        public Task Release(string category, AcquiredLease[] acquiredLeases)
+        {
+            foreach(AcquiredLease lease in acquiredLeases)
+            {
+                Release(category, lease);
+            }
+            return Task.CompletedTask;
+        }
+
+        public Task<AcquireLeaseResult[]> Renew(string category, AcquiredLease[] acquiredLeases)
+        {
+            return Task.FromResult(acquiredLeases.Select(lease => Renew(category, lease)).ToArray());
+        }
+
+        private AcquireLeaseResult Acquire(string category, LeaseRequest leaseRequest)
+        {
+            DateTime now = DateTime.UtcNow;
+            Lease lease = this.leases.GetValueOrAddNew(Tuple.Create(category, leaseRequest.ResourceKey));
+            if(lease.ExpiredUtc < now)
+            {
+                lease.ExpiredUtc = now + leaseRequest.Duration;
+                return new AcquireLeaseResult(new AcquiredLease(leaseRequest.ResourceKey, leaseRequest.Duration, lease.Token, now), ResponseCode.OK, null);
+            }
+            return new AcquireLeaseResult(null, ResponseCode.LeaseNotAvailable, new OrleansException("Lease not available"));
+        }
+
+        private void Release(string category, AcquiredLease acquiredLease)
+        {
+            Tuple<string,string> leaseKey = Tuple.Create(category, acquiredLease.ResourceKey);
+            if (this.leases.TryGetValue(leaseKey, out Lease lease) && lease.Token == acquiredLease.Token)
+            {
+                leases.Remove(leaseKey);
+            }
+        }
+
+        private AcquireLeaseResult Renew(string category, AcquiredLease acquiredLease)
+        {
+            DateTime now = DateTime.UtcNow;
+            // if lease exists, and we have the right token, and lease has not expired, renew.
+            if (this.leases.TryGetValue(Tuple.Create(category, acquiredLease.ResourceKey), out Lease lease))
+            {
+                // check token
+                if (lease.Token != acquiredLease.Token)
+                {
+                    return new AcquireLeaseResult(null, ResponseCode.InvalidToken, new OrleansException("Invalid token provided, caller is not the owner."));
+                }
+                // we don't care if lease has expired or not as long was owner has not changed.
+                lease.ExpiredUtc = now + acquiredLease.Duration;
+                return new AcquireLeaseResult(new AcquiredLease(acquiredLease.ResourceKey, acquiredLease.Duration, lease.Token, now), ResponseCode.OK, null);
+            }
+            // else try an aquire
+            return Acquire(category, new LeaseRequest { ResourceKey = acquiredLease.ResourceKey, Duration = acquiredLease.Duration } );
```

Comment:
No objections to changing this, but I'd like to understand the expected behavior first.

- If renewing a lease with a valid token and lease has not expired, renew.
- If renewing a lease with an invalid token, invalid token error.

However:
- **If renewing a lease with a valid tag, but lease has expired, what do we do?**  In this code I'm renewing it anyway.
- **If renewing a lease on a resource that is owned by no one, (not even the caller) what should we do?**  In this code, I'm treating the renew as an acquire.

I selected the above two behaviors because they are best attempts at servicing the callers request and do no harm.